### DISCUSSION TITLE
How to fill a PyTorch string_view input

### DISCUSSION THREAD

**Discussion by sbrunk** [bytedeco/javacpp-presets#1390] [Q&A] [answered]
The PyTorch bindings have a bunch of functions taking a `c10:string_view`, which is currently mapped to a Pointer.

Example:

```java
@Namespace("at") public static native @ByVal Tensor einsum(@ByVal @Cast("c10::string_view*") Pointer equation, @ByVal @Cast("at::TensorList*") TensorArrayRef tensors);
```

I'm trying to call `einsum` with a string. My naive approach was to just initialize a `BytePointer`

```java
import org.bytedeco.pytorch.*;
var rand = global.torch.torch_randn(4,4);
global.torch.einsum(BytePointer("ii"), new TensorArrayRef(rand.position(0), 1));
```

But that doesn't work:

```
Caused by: java.lang.RuntimeException: einsum(): the number of subscripts in the equation (0) does not match the number of dimensions (2) for operand 0 and no ellipsis was given
```

The error is the same when calling the Python API with an empty string.

Any idea how to correctly add the `string_view` value or does that need mapping support (i.e. a cast)?

Tagging @HGuillemet because you're deep into the PyTorch mappings right now.

**Comment by HGuillemet** [comment]
I'll look into this.
I think we could add a mapping.

**Comment by HGuillemet** [comment]
@saudet, please advise about this:

I first tried to create a custom adapter `StringViewAdapter` implementing proper conversions, but it causes problem, that I could explicit if needed, when instantiating containers like `c10::optional<c10::string_view>`.

Then I realized that the `@StdString` adapter is almost sufficient.
For functions returning a `string_view`, we should generate:
```Java
@Cast({"", "", "std::basic_string<char>"}) @StdString String f()
```
because string_view knows how to be cast to `std::basic_string` only.

For functions taking a string_view, we simply need:
```
void f(@StdString String)
```
We must not cast to `std::basic_string`. string_view has an implicit constructor taking a `char *`.

Is there an existing trick to generate the `@Cast` only for returned values ?
Else should we consider adding this option to Parser ?

A third option that should work, but less friendly, is to map string_view to a specific Java class.

**Reply by saudet** [reply]
It's probably easier to come up with something like a StringViewAdapter, possibly based on StringAdapter like this:
https://github.com/bytedeco/javacpp-presets/blob/master/tensorflow/src/main/resources/org/bytedeco/tensorflow/include/tensorflow_adapters.h#L53

**Reply by HGuillemet** [reply, marked as answer]
Ok. I have came up with a solution with an adapter working. New PR coming.
Like in `StringAdapter`, I tested the size passed to the `(ptr, size, owner)` constructors and did a `strlen` if it's 0. 
But I wonder if it won't be better to have the generator produce a JNI call to `getStringUTFLength`, along with `getStringUTFChars`, and pass this length to the adapter.

**Reply by sbrunk** [reply]
For reference: The PR is https://github.com/bytedeco/javacpp-presets/pull/1392 and since it's merged, this is now working fine.
